action/protocol/staking/vote_reviser.go

```diff
@@ -2,38 +2,54 @@ package staking
 
 import (
 	"math/big"
+	"slices"
 	"sort"
 
 	"github.com/pkg/errors"
 	"go.uber.org/zap"
 
+	"github.com/iotexproject/iotex-core/action/protocol"
 	"github.com/iotexproject/iotex-core/blockchain/genesis"
 	"github.com/iotexproject/iotex-core/pkg/log"
 	"github.com/iotexproject/iotex-core/state"
 )
 
 // VoteReviser is used to recalculate candidate votes.
 type VoteReviser struct {
-	reviseHeights      []uint64
-	cache              map[uint64]CandidateList
-	c                  genesis.VoteWeightCalConsts
-	correctCandsHeight uint64
+	reviseHeights           []uint64
+	cache                   map[uint64]CandidateList
+	c                       genesis.VoteWeightCalConsts
+	correctCandsHeight      uint64
+	correctSelfStakeHeights []uint64
 }
 
 // NewVoteReviser creates a VoteReviser.
-func NewVoteReviser(c genesis.VoteWeightCalConsts, correctCandsHeight uint64, reviseHeights ...uint64) *VoteReviser {
+func NewVoteReviser(c genesis.VoteWeightCalConsts, correctCandsHeight uint64, reviseHeights []uint64, correctSelfStakeHeights []uint64) *VoteReviser {
```

Comment:
also need to update logic in `NeedRevise`?

Reply:
add correctSelfStake and correctEndorsement